jormungandr/src/blockchain/storage.rs

```diff
@@ -180,105 +60,109 @@ enum StreamingError {
+        }
     }
 
-    pub async fn block_exists(&self, header_hash: HeaderHash) -> Result<bool, Error> {
-        self.run(
-            move |connection| match connection.block_exists(&header_hash) {
-                Err(StorageError::BlockNotFound) => Ok(false),
-                Ok(r) => Ok(r),
-                Err(e) => Err(e),
-            },
-        )
-        .await
+    pub fn block_exists(&self, header_hash: &HeaderHash) -> Result<bool, Error> {
+        self.storage
+            .block_exists(header_hash.as_ref())
+            .map_err(Into::into)
     }
 
-    pub async fn get_blocks_by_chain_length(&self, chain_length: u64) -> Result<Vec<Block>, Error> {
-        self.run(
-            move |connection| match connection.get_blocks_by_chain_length(chain_length) {
-                Err(StorageError::BlockNotFound) => Ok(Vec::new()),
-                Ok(r) => Ok(r.into_iter().map(|(block, _)| block).collect()),
-                Err(e) => Err(e),
-            },
-        )
-        .await
+    pub fn get_blocks_by_chain_length(&self, chain_length: u32) -> Result<Vec<Block>, Error> {
+        self.storage
+            .get_blocks_by_chain_length(chain_length)
+            .map_err(Into::into)
+            .and_then(|blocks| {
+                blocks
+                    .into_iter()
+                    .map(|block| Block::deserialize(block.as_ref()))
+                    .collect::<Result<Vec<_>, _>>()
+                    .map_err(Error::Deserialize)
+            })
     }
 
-    pub async fn put_block(&self, block: Block) -> Result<(), Error> {
-        self.run(move |connection| match connection.put_block(&block) {
-            Err(StorageError::BlockNotFound) => unreachable!(),
-            Err(e) => Err(e),
-            Ok(()) => Ok(()),
-        })
-        .await
+    pub fn put_block(&self, block: &Block) -> Result<(), Error> {
+        let id = block
+            .header
+            .hash()
+            .serialize_as_vec()
+            .map_err(Error::Serialize)?;
+        let parent_id = block
+            .header
+            .block_parent_hash()
+            .serialize_as_vec()
+            .map_err(Error::Serialize)?;
+        let chain_length = block.header.chain_length().into();
+        let block_info = BlockInfo::new(id, parent_id, chain_length);
+        self.storage
+            .put_block(
+                &block.serialize_as_vec().map_err(Error::Serialize)?[..],
+                block_info,
+            )
+            .map_err(Into::into)
     }
 
     /// Return values:
     /// - `Ok(stream)` - `from` is ancestor of `to`, returns blocks between them
     /// - `Err(CannotIterate)` - `from` is not ancestor of `to`
     /// - `Err(BlockNotFound)` - `from` or `to` was not found
     /// - `Err(_)` - some other storage error
-    pub async fn stream_from_to(
+    pub fn stream_from_to(
         &self,
-        from: HeaderHash,
-        to: HeaderHash,
+        from: &HeaderHash,
+        to: &HeaderHash,
```

Comment:
`HeaderHash` is `Copy` and a small fixed array, it's not necessary to pass these values by reference.